test/dtypes/test_int4.py

```diff
@@ -0,0 +1,321 @@
+        fake_in_features, fake_out_features = 8, 8
+        new_mod = cls(
+            fake_in_features,
+            fake_out_features,
+            bias=mod.bias is not None,
+            w_int4=w_int4.t().contiguous(),
+            scales=scales,
+        )
+        new_mod.in_features = mod.in_features
+        new_mod.out_features = mod.out_features
+        del new_mod.weight
+        new_mod.bias = mod.bias
+        device_to_use = next(mod.parameters()).device
+        new_mod.to(device_to_use)
+        return new_mod
+
+def _apply_weight_only_int4_quant(model):
+    replace_with_custom_fn_if_matches_filter(
+        model,
+        _WeightOnlyInt4QuantLinear.from_float,
+        lambda mod, fqn: isinstance(mod, torch.nn.Linear),
+    )
+
+from torch.library import Library, impl
+
+test_lib = Library("test_int4", "DEF")
+test_lib.define("quantize_per_tensor_int4(Tensor input, float scale, int zero_point) -> Tensor")
+
+@impl(test_lib, "quantize_per_tensor_int4", "CompositeExplicitAutograd")
+def quantize_per_tensor_int4(
+    input: torch.Tensor,
+    scale: float,
+    zero_point: int,
+) -> torch.Tensor:
+    inv_scale = 1.0 / scale
+    return torch.clamp(torch.round(input * inv_scale) + zero_point, 0, 15).to(torch.uint8).view(torch.bits8)
+
+test_lib.define("dequantize_per_tensor_int4(Tensor input, float scale, int zero_point) -> Tensor")
+@impl(test_lib, "dequantize_per_tensor_int4", "CompositeExplicitAutograd")
+def dequantize_per_tensor_int4(
+    input: torch.Tensor,
+    scale: float,
+    zero_point: int,
+) -> torch.Tensor:
+    return (input.view(torch.uint8).to(torch.float32) - zero_point) * scale
+
+
+class TestInt4(QuantizationTestCase):
+    def test_basic_tensor_ops(self):
+        x = UInt4Tensor(torch.tensor([
+            [0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF],
+            [0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF],
+            [0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF],
+        ], dtype=torch.uint8))
+        self.assertTrue(x.shape, (3, 8))
+        # making sure these works
+        x.to(torch.uint8)
+        expected = UInt4Tensor(torch.tensor([
+            [0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF],
+        ], dtype=torch.uint8))
+        self.assertTrue(x[0:1, :] == expected)
+        expected = UInt4Tensor(torch.tensor([
+            [0x23, 0x45],
+            [0x23, 0x45],
+            [0x23, 0x45],
+        ], dtype=torch.uint8))
+        self.assertTrue(x[:, 2:6] == expected)
+
+    def test_gpu_quant(self):
+        for x_shape in [[2, 4], [5, 5, 5, 4], [1, 4, 4]]:
+            x = torch.randn(*x_shape)
+            m = nn.Sequential(nn.Linear(4, 16))
+            y_ref = m(x)
+            _apply_weight_only_int4_quant(m)
+            y_wo = m(x)
+            # sqnr = compute_error(y_ref, y_wo)
+            opt = torch.compile(m, mode="max-autotune")
+            # make sure it runs
+            opt(x)
```

Comment:
I'm seeing an assertion failure [here](https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/meta_utils.py#L474)

When I look into it more:

(1) `symbolic_context` is a `StatelessSymbolicContext`, not a `SubclassSymbolicContext`

(2) that's because when we fakeify a tensor that is a view, we unconditionally create a `StatelessSymbolicContext` for its base: https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/meta_utils.py#L330

(3) However in this case, the ._base of the tensor we're fakifying is another subclass, so it should get a `SubclassSymbolicContext`

I'm going to make a min repro and file an issue for it. Joel has also been doing some work around fakifying subclasses that are views, so hopefully his work will also fix this bug.

Reply:
@jerryzh168 this patch should also avoid the issue for you:
```
--- a/test/dtypes/test_int4.py
+++ b/test/dtypes/test_int4.py
@@ -91,7 +91,7 @@ class _WeightOnlyInt4QuantLinear(torch.nn.Linear):
             fake_in_features,
             fake_out_features,
             bias=mod.bias is not None,
-            w_int4=w_int4.t().contiguous(),
+            w_int4=torch.ops.aten.transpose_copy.default(w_int4),
             scales=scales,
         )
         new_mod.in_features = mod.in_features
```

It looks like the way that `UInt4Tensor.__torch_dispatch__` implements transpose, it always returns a contiguous tensor (is that what you want?). This means that:

(1) `w_int4.t()` will report as being contiguous (`w_int4.t().is_contiguous() == True`)

(2) the `.contiguous()` call afterwards will be a no-op

(3) the autograd engine will think that `w_int4.t().contiguous()` is a view, and not a fresh tensor. This means that the final weight in your nn module, `w_int4`, will have a `._base` field, causing it to go down a more interesting path in the fakeification logic.

One easy workaround for now is that since you know that your subclass implements transpose to return a contiguous tensor, you can just call `transpose_copy` from the user code, so autograd will know that the output does not alias the original tensor.

Another option though is that you might want to implement `transpose` with the proper semantics, so that it returns a tensor that is not contiguous.